src/main/java/com/github/kornilova_l/flamegraph/plugin/server/converters/file/jmc/FlightRecorderCompressedConverter.java

```diff
@@ -0,0 +1,87 @@
+package com.github.kornilova_l.flamegraph.plugin.server.converters.file.jmc;
+
+import com.github.kornilova_l.flamegraph.plugin.server.converters.file.CFlamegraph;
+import com.github.kornilova_l.flamegraph.plugin.server.converters.file.CFlamegraphLine;
+
+import java.io.File;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Set;
+
+class FlightRecorderCompressedConverter extends FlightRecorderConverter {
+    private CFlamegraph cFlamegraph;
+
+    public FlightRecorderCompressedConverter(File file) {
+        super(file);
+
+        parseAsCFlamegraph();
+    }
+
+    CFlamegraph getCFlamegraph() {
+        return cFlamegraph;
+    }
+
+    private void parseAsCFlamegraph() {
```

Comment:
This is the tricky part. I could not get it from the first time either. Here is my solution that is not efficient and contains a bug.
Hope, after couple of dev iterations we will get it right 😄 
```java
private void parseAsCFlamegraph() {
        ArrayList<CFlamegraphLine> cFlamegraphLines = new ArrayList<>();
        HashMap<String, Integer> classNames = new HashMap<>();
        HashMap<String, Integer> methodNames = new HashMap<>();
        HashMap<String, Integer> descriptions = new HashMap<>();

        TreeMap<String, Integer> sortedStacks = new TreeMap<>(this.getStacks());
        // todo: it's better to split stacks beforehand.
        //  Currently split(";") is called n^2 times (in for loop and in getWidth method)
        //  Maybe something like: List<Stack> sortedStacks
        //  class Stack { String[] frames; int width }
        //  also it's good to filter here stacks with 0 width
        for (Map.Entry<String, Integer> stack : sortedStacks.entrySet()) {
            String[] frames = stack.getKey().split(";");
            boolean skippingDuplicates = true;
            for (int depth = 1; depth <= frames.length; depth++) {
                // I don't remember what was the problem about depth = 0
                // probably something related to serialization. So it should start with 1
                String frame = frames[depth - 1];
                String className = getClassName(frame);
                int classNameId = getId(classNames, className);
                String methodName = getMethodName(frame);
                int methodNameId = getId(methodNames, methodName);
                if (skippingDuplicates && lastNodeIsTheSame(cFlamegraphLines, classNameId, methodNameId, depth)) continue;
                skippingDuplicates = false;
                int width = getWidth(sortedStacks, frame, depth);
                if (width == 0) continue;
                cFlamegraphLines.add(new CFlamegraphLine(
                        classNameId,
                        methodNameId,
                        null,
                        width,
                        depth
                ));
            }
        }
        this.cFlamegraph = new CFlamegraph(
                cFlamegraphLines,
                toArray(classNames),
                toArray(methodNames),
                toArray(descriptions)
        );
    }

    private boolean lastNodeIsTheSame(ArrayList<CFlamegraphLine> cFlamegraphLines, int classNameId, int methodNameId, int depth) {
        if (cFlamegraphLines.isEmpty()) return false;
        // todo: check description
        for (int i = cFlamegraphLines.size() - 1; i >= 0; i--) {
            CFlamegraphLine cFlamegraphLine = cFlamegraphLines.get(i);
            if (cFlamegraphLine.getDepth() == depth) {
                Integer clNameId = cFlamegraphLine.getClassNameId();
                return clNameId != null && clNameId == classNameId && cFlamegraphLine.getMethodNameId() == methodNameId;
            }
        }
        return false;
    }

    private int getWidth(Map<String, Integer> stacks, String frame, int depth) {
        // todo: bug here.
        //  We need to calc width across all neighbouring frames on the same depth
        //  (not across all frames on the same depth).
        //  This can be fixed after we solve above problem ^^ about splitting stacks beforehand.
        int width = 0;
        for (Map.Entry<String, Integer> stack : stacks.entrySet()) {
            String[] frames = stack.getKey().split(";");
            if (depth > frames.length) continue;
            String currentFrame = frames[depth - 1];
            if (currentFrame.equals(frame)) width += stack.getValue();
        }
        return width;
    }

```

Reply:
I don't fully understand the intent of this code, so I'm having a hard time determining how to optimize. I feel like the data structures we use to handle splitting `;` only once will impact the width aggregation of neighboring frames. 

It looks like the final display is driven using `TreesProtos.Tree`, which is built and serialized by the generated CFlamegraph in `CompressedFlamegraphFileSave.tryToConvert` (which goes CFlamegraph -> TreesProto.Tree, saved to a file). The following seems to handle what we do in your above code:

`TreeProtos.Tree tree = new StacksToTreeBuilder(this.getStacks()).getTree();`

We'd need to convert `TreesProtos.Tree` to `CFlamegraph`. This would look like `CompressedFlamegraphFileSave.tryToConvert` but in the opposite direction.

Am I looking at this wrong?

Reply:
You are confused because there are two `Tree` classes:
1. `cflamegraph.Tree` flatbuffers class that is used to store compressed flamegraph on disk
2. `proto.TreeProtos.Tree` google protobuf class that is used for visualization. It has different structure and contains much more information than `cflamegraph.Tree`, so it's easy to draw it in browser.

So `CompressedFlamegraphFileSave.tryToConvert` converts everything to `cflamegraph.Tree` and `CompressedFlamegraphToCallTracesConverter` converts `cflamegraph.Tree` to `proto.TreeProtos.Tree`.

In fact `StacksToTreeBuilder` is not used here. Take a look at `cflamegraph.Converter` it parses cflamegraph and directly builds `proto.TreeProtos.Tree`